Arm/Arm.cs

```diff
@@ -0,0 +1,290 @@
+         *   the arm will ensure that that ArmPart will always be at
+         *   that angle with respect to the ground. 
+         */
+
+        public float MinRoll;
+        public float MaxRoll;
+        public float MinPitch;
+        public float MaxPitch;
+        public float MinYaw;
+        public float MaxYaw;
+        public float Length;
+
+        public bool Fixed;
+        public float FixedAngle;
+
+        public ArmPart(float MinRoll, float MaxRoll,
+                       float MinPitch, float MaxPitch,
+                       float MinYaw, float MaxYaw,
+                       float L)
+        {
+            this.MinRoll = MinRoll;
+            this.MaxRoll = MaxRoll;
+            this.MinPitch = MinPitch;
+            this.MaxPitch = MaxPitch;
+            this.MinYaw = MinYaw;
+            this.MaxYaw = MaxYaw;
+            Length = L;
+            Fixed = false;
+            FixedAngle = 0.0f;
+        }
+
+        public ArmPart(float FixedAngle, float Length)
+        {
+            this.MinRoll = 0.0f;
+            this.MaxRoll = 0.0f;
+            this.MinPitch = 0.0f;
+            this.MaxPitch = (float)Math.PI * 2.0f;
+            this.MinYaw = 0.0f;
+            this.MaxYaw = 0.0f;
+            this.FixedAngle = FixedAngle;
+            this.Fixed = true;
+            this.Length = Length;
+        }
+
+        public static implicit operator ArmPart((int, int) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2);
+        }
+
+        public static implicit operator ArmPart((float, float) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2);
+        }
+
+        public static implicit operator ArmPart((double, double) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2);
+        }
+
+        public static implicit operator ArmPart((int, int, int, int, int, int, int) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2, (float)t.Item3, (float)t.Item4, (float)t.Item5,
+                             (float)t.Item6, (float)t.Item7);
+        }
+
+        public static implicit operator ArmPart((double, double, double, double, double, double, double) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2, (float)t.Item3, (float)t.Item4, (float)t.Item5,
+                  (float)t.Item6, (float)t.Item7);
+        }
+
+        public static implicit operator ArmPart((float, float, float, float, float, float, float) t)
+        {
+            return new ArmPart((float)t.Item1, (float)t.Item2, (float)t.Item3, (float)t.Item4, (float)t.Item5,
+                  (float)t.Item6, (float)t.Item7);
+        }
+    }
+
+    public class Arm
+    {
+        ArmPart[] Params;
+        public float[] CurrentPitches { get; private set; }
+        public float[] CurrentRolls { get; private set; }
+        public float[] CurrentYaws { get; private set; }
+
+        private float[] TempPitches;
```

Comment:
We need this because we need simultaneous updating of angles. Allocating an array inside the function call is still allocate on the heap. This way we only have to allocate once in constructor.

Reply:
If we want to allocate the temporary arrays on the stack, we could allow unsafe code and use the stackalloc keyword. Probably not what we want to do.